packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,338 @@
+    );
+
+    this.libp2p.handle(FilterV2Codecs.PUSH, this.onRequest.bind(this));
+
+    this.options = options ?? {};
+  }
+
+  async createSubscription(
+    _pubSubTopic?: string,
+    peerId?: PeerId
+  ): Promise<Subscription> {
+    const pubSubTopic =
+      _pubSubTopic ?? this.options.pubSubTopic ?? DefaultPubSubTopic;
+
+    const peer = await this.getPeer(peerId);
+
+    return new Subscription(pubSubTopic, peer, this.newStream.bind(this, peer));
+  }
+
+  public toSubscriptionIterator<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    opts?: ProtocolOptions | undefined
+  ): Promise<IAsyncIterator<T>> {
+    return toAsyncIterator(this, decoders, opts);
+  }
+
+  private onRequest(streamData: IncomingStreamData): void {
+    log("Receiving message push");
+    try {
+      pipe(streamData.stream, lp.decode, async (source) => {
+        for await (const bytes of source) {
+          const response = FilterPushRpc.decode(bytes.slice());
+
+          const { pubsubTopic, wakuMessage } = response;
+
+          if (!wakuMessage) {
+            log("Received empty message");
+            return;
+          }
+
+          const subs = FilterV2.subscriptionsLog as Map<
+            PeerIdStr,
+            PeerSubscription<IDecodedMessage>[]
+          >;
+          const subscription = subs
+            .get(streamData.connection.remotePeer.toString())
+            ?.find((s) => s.pubSubTopic === pubsubTopic);
+
+          if (!subscription) {
+            log(`No subscription locally registered for topic ${pubsubTopic}`);
+            return;
+          }
+
+          await this.pushMessage(subscription, wakuMessage);
+        }
+      }).then(
+        () => {
+          log("Receiving pipe closed.");
+        },
+        (e) => {
+          log("Error with receiving pipe", e);
+        }
+      );
+    } catch (e) {
+      log("Error decoding message", e);
+    }
+  }
+
+  private async pushMessage<T extends IDecodedMessage>(
+    peerSubscription: PeerSubscription<T>,
+    message: WakuMessage
+  ): Promise<void> {
+    const { decoders, callback, pubSubTopic } = peerSubscription;
+
+    const { contentTopic } = message;
+    if (!contentTopic) {
+      log("Message has no content topic, skipping");
+      return;
+    }
+
+    let didDecodeMsg = false;
+    // We don't want to wait for decoding failure, just attempt to decode
+    // all messages and do the call back on the one that works
+    // noinspection ES6MissingAwait
+    decoders.forEach(async (dec: IDecoder<T>) => {
+      if (didDecodeMsg) return;
+      const decoded = await dec.fromProtoObj(
+        pubSubTopic,
+        message as IProtoMessage
+      );
+      // const decoded = await dec.fromProtoObj(pubSubTopic, message);
+      if (!decoded) {
+        log("Not able to decode message");
+        return;
+      }
+      // This is just to prevent more decoding attempt
+      // TODO: Could be better if we were to abort promises
```

Comment:
Could you create an issue to track? This is a refactor pattern present in several places so having one issue that can re-implement the pattern everywhere at once would be good. in short, we need to use abortable promises.